tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc

```diff
@@ -57,5 +97,113 @@ void CustomGraphOptimizerRegistry::RegisterOptimizerOrDie(
   GetRegistrationMap()->insert({name, optimizer_creator});
 }
 
+std::vector<std::unique_ptr<CustomGraphOptimizer>>
+PluginGraphOptimizerRegistry::CreateOptimizer(
+    const std::set<string>& device_types) {
+  std::vector<std::unique_ptr<CustomGraphOptimizer>> optimizer_list;
+  for (auto it = GetPluginRegistrationMap()->begin();
+       it != GetPluginRegistrationMap()->end(); it++) {
+    if (device_types.find(it->first) == device_types.end()) continue;
+    LOG(INFO) << "Plugin optimizer for device_type " << it->first
+              << " is enabled.";
+    optimizer_list.emplace_back(
+        std::unique_ptr<CustomGraphOptimizer>(it->second()));
+  }
+  return optimizer_list;
+}
+
+void PluginGraphOptimizerRegistry::RegisterPluginOptimizerOrDie(
+    const Creator& optimizer_creator, const std::string& device_type,
+    ConfigsList& configs) {
+  auto ret = GetPluginConfigMap()->insert({device_type, configs});
+  if (!ret.second) {
+    LOG(FATAL) << "PluginGraphOptimizer with device_type " << device_type
```

Comment:
We don't allow `LOG(FATAL)` in new code. Can we just make TF warn that it finds a duplicate (which TF will ignore) and carry on without crashing? (Let's also remove `OrDie` from the function name.) If something can go wrong with this route, please explain why.

Reply:
Taking back my comment. We agree it's okay to fail. Please ignore this. :)